File: task 2/TicTacToe.java
import java.util.Scanner;

public class TicTacToe {

    private static char[] board = {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '};
    private static char player = 'X';
    private static char ai = 'O';

    public static void printBoard() {
        System.out.println();
        for (int i = 0; i < 9; i += 3) {
            System.out.println(" " + board[i] + " | " + board[i + 1] + " | " + board[i + 2] + " ");
            if (i < 6) {
                System.out.println("---|---|---");
            }
        }
        System.out.println();
    }

    private static boolean isWinner(char[] bo, char le) {
        return (bo[0] == le && bo[1] == le && bo[2] == le) ||
               (bo[3] == le && bo[4] == le && bo[5] == le) ||
               (bo[6] == le && bo[7] == le && bo[8] == le) ||
               (bo[0] == le && bo[3] == le && bo[6] == le) ||
               (bo[1] == le && bo[4] == le && bo[7] == le) ||
               (bo[2] == le && bo[5] == le && bo[8] == le) ||
               (bo[0] == le && bo[4] == le && bo[8] == le) ||
               (bo[2] == le && bo[4] == le && bo[6] == le);
    }

    private static boolean isBoardFull(char[] bo) {
        for (char c : bo) {
            if (c == ' ') {
                return false;
            }
        }
        return true;
    }

    private static int minimax(char[] newBoard, boolean isMaximizing) {
        if (isWinner(newBoard, ai)) {
            return 1;
        } else if (isWinner(newBoard, player)) {
            return -1;
        } else if (isBoardFull(newBoard)) {
            return 0;
        }

        if (isMaximizing) {
            int bestScore = Integer.MIN_VALUE;
            for (int i = 0; i < 9; i++) {
                if (newBoard[i] == ' ') {
                    newBoard[i] = ai;
                    int score = minimax(newBoard, false);
                    newBoard[i] = ' ';
                    bestScore = Math.max(score, bestScore);
                }
            }
            return bestScore;
        } else {
            int bestScore = Integer.MAX_VALUE;
            for (int i = 0; i < 9; i++) {
                if (newBoard[i] == ' ') {
                    newBoard[i] = player;
                    int score = minimax(newBoard, true);
                    newBoard[i] = ' ';
                    bestScore = Math.min(score, bestScore);
                }
            }
            return bestScore;
        }
    }

    private static int getBestMove() {
        int bestScore = Integer.MIN_VALUE;
        int move = 0;
        for (int i = 0; i < 9; i++) {
            if (board[i] == ' ') {
                board[i] = ai;
                int score = minimax(board, false);
                board[i] = ' ';
                if (score > bestScore) {
                    bestScore = score;
                    move = i;
                }
            }
        }
        return move;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Welcome to Tic-Tac-Toe!");
        System.out.println("You are X, and the AI is O.");
        printBoard();

        while (true) {
            // Player's move
            int playerMove;
            while (true) {
                System.out.print("Enter your move (1-9): ");
                try {
                    playerMove = scanner.nextInt() - 1;
                    if (playerMove >= 0 && playerMove < 9 && board[playerMove] == ' ') {
                        board[playerMove] = player;
                        break;
                    } else {
                        System.out.println("Invalid move. Try again.");
                    }
                } catch (Exception e) {
                    System.out.println("Invalid input. Please enter a number between 1 and 9.");
                    scanner.next(); // clear the invalid input
                }
            }
            printBoard();

            if (isWinner(board, player)) {
                System.out.println("Congratulations! You won!");
                break;
            } else if (isBoardFull(board)) {
                System.out.println("It's a draw!");
                break;
            }

            // AI's move
            System.out.println("AI is thinking...");
            int aiMove = getBestMove();
            board[aiMove] = ai;
            System.out.println("AI chose position " + (aiMove + 1));
            printBoard();

            if (isWinner(board, ai)) {
                System.out.println("AI wins! Better luck next time.");
                break;
            } else if (isBoardFull(board)) {
                System.out.println("It's a draw!");
                break;
            }
        }
        scanner.close();
    }
}

File: task 2/TicTacToeAI/tictactoe.py
import math

# Represents the game board
board = [' ' for _ in range(9)]

def print_board():
    """Prints the Tic-Tac-Toe board."""
    print()
    for i in range(0, 9, 3):
        print(f" {board[i]} | {board[i+1]} | {board[i+2]} ")
        if i < 6:
            print("---|---|---")
    print()

def is_winner(bo, le):
    """Checks if a player has won."""
    return ((bo[0] == le and bo[1] == le and bo[2] == le) or
            (bo[3] == le and bo[4] == le and bo[5] == le) or
            (bo[6] == le and bo[7] == le and bo[8] == le) or
            (bo[0] == le and bo[3] == le and bo[6] == le) or
            (bo[1] == le and bo[4] == le and bo[7] == le) or
            (bo[2] == le and bo[5] == le and bo[8] == le) or
            (bo[0] == le and bo[4] == le and bo[8] == le) or
            (bo[2] == le and bo[4] == le and bo[6] == le))

def is_board_full(bo):
    """Checks if the board is full."""
    return ' ' not in bo

def minimax(bo, depth, is_maximizing):
    """Minimax algorithm implementation."""
    if is_winner(bo, 'O'):
        return 1
    elif is_winner(bo, 'X'):
        return -1
    elif is_board_full(bo):
        return 0

    if is_maximizing:
        best_score = -math.inf
        for i in range(9):
            if bo[i] == ' ':
                bo[i] = 'O'
                score = minimax(bo, depth + 1, False)
                bo[i] = ' '
                best_score = max(score, best_score)
        return best_score
    else:
        best_score = math.inf
        for i in range(9):
            if bo[i] == ' ':
                bo[i] = 'X'
                score = minimax(bo, depth + 1, True)
                bo[i] = ' '
                best_score = min(score, best_score)
        return best_score

def get_best_move():
    """Finds the best move for the AI."""
    best_score = -math.inf
    move = 0
    for i in range(9):
        if board[i] == ' ':
            board[i] = 'O'
            score = minimax(board, 0, False)
            board[i] = ' '
            if score > best_score:
                best_score = score
                move = i
    return move

def main():
    """Main game loop."""
    print("Welcome to Tic-Tac-Toe!")
    print("You are X, and the AI is O.")
    print_board()

    while True:
        # Player's move
        while True:
            try:
                player_move = int(input("Enter your move (1-9): ")) - 1
                if 0 <= player_move <= 8 and board[player_move] == ' ':
                    board[player_move] = 'X'
                    break
                else:
                    print("Invalid move. Try again.")
            except ValueError:
                print("Invalid input. Please enter a number between 1 and 9.")

        print_board()

        if is_winner(board, 'X'):
            print("Congratulations! You won!")
            break
        elif is_board_full(board):
            print("It's a draw!")
            break

        # AI's move
        print("AI is thinking...")
        ai_move = get_best_move()
        board[ai_move] = 'O'
        print(f"AI chose position {ai_move + 1}")
        print_board()

        if is_winner(board, 'O'):
            print("AI wins! Better luck next time.")
            break
        elif is_board_full(board):
            print("It's a draw!")
            break

if __name__ == "__main__":
    main()
